### PR TITLE
[Feature] `TensorDict.split` method

### PR DESCRIPTION
**Pull request by wonnor-pro**
## Description

- Implement `tensordict.split` method that works like [tensor.split](https://pytorch.org/docs/stable/generated/torch.split.html).
- Overload `torch.split` to support tensordict.
- Add test cases in `test_tensordict.py`

## Motivation and Context

As requested in issue #28, It would be nice to have a tensordict.split method that would work as [tensor.split](https://pytorch.org/docs/stable/generated/torch.split.html).
We would also need to overload torch.split as we do for torch.stack if possible.

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [x] New feature (non-breaking change which adds core functionality)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [x] I have updated the documentation accordingly.

My changes involve a new function in API, the docstring is added to the function. The auto-generated file is not included in the commit.

### CODE REVIEW

File: tensordict/tensordict.py [hunk 1]
```diff
@@ -1424,6 +1424,36 @@ def reshape(
             batch_size = shape
         return TensorDict(d, batch_size, device=self.device)
 
+    def split(
+        self, split_size: Union[int, List[int]], dim: int = 0
+    ) -> List[TensorDictBase]:
+        """Splits each tensor in the TensorDict with the specified size in the given dimension, like `torch.split`.
+        Returns a list of TensorDict with the view of split chunks of tensor.
```
Comment (vmoens) [inline tensordict/tensordict.py:1431]: I would add an empty line before this

File: tensordict/tensordict.py [hunk 2]
```diff
@@ -1424,6 +1424,36 @@ def reshape(
...
+
+        The list of TensorDict maintains the original order of the tensor chunks.
+
+        Args:
+            split_size (int) or (list(int)): size of a single chunk or list of sizes for each chunk
```
Comment (vmoens) [inline tensordict/tensordict.py:1436]: `split_size (int of List[int]): ...`

File: tensordict/tensordict.py [hunk 3]
```diff
@@ -1424,6 +1424,36 @@ def reshape(
...
+            A list of TensorDict with specified size in given dimension.
+
+        """
+        dictionaries = []
+        batch_size = []
+        for key, item in self.items():
+            split_tensors = torch.split(item, split_size, dim)
+            if not dictionaries:
+                dictionaries = [dict() for _ in range(len(split_tensors))]
+                batch_size = [split_tensor.shape[:-1] for split_tensor in split_tensors]
```
Comment (vmoens) [inline tensordict/tensordict.py:1449]: I'm a bit confused by the batch size

If you have this case

```
t = TensorDict({}, [3, 10, 5])
t.split([3, 5, 2], 1)
```
you would get 3 (empty) tensordicts of batch size `[3, 3, 5]`, `[3, 5, 5]` and`[3, 2, 5]`.
here that would not be the result i think

File: test/test_tensordict.py [hunk 4]
```diff
@@ -2935,6 +2935,52 @@ def test_unflatten_keys_collision(separator):
...
+    assert tds[1].shape == torch.Size([5])
+
+    tds = td.split([5, 2, 3], 0)
+    assert len(tds) == 3
+    assert tds[0].shape == torch.Size([5])
+    assert tds[1].shape == torch.Size([2])
+    assert tds[2].shape == torch.Size([3])
+
+    td = TensorDict(
+        source={"a": torch.zeros(5, 6, 5), "b": torch.zeros(5, 6, 10, 1)},
```
Comment (vmoens) [inline test/test_tensordict.py:2953]: I guess that if you swap the values of a and b, since you take the shape of the first tensor up to the last-but-one dimension, you'll end up with a batch size that is not accurate no?

File: test/test_tensordict.py [hunk 5]
```diff
@@ -2935,6 +2935,52 @@ def test_unflatten_keys_collision(separator):
     assert assert_allclose_td(td.unflatten_keys(separator), ref)
 
 
+def test_split():
```
Comment (vmoens) [inline test/test_tensordict.py:2938]: Another version of this test would be to test with this setting:

Put the following tensordicts in a `pytest.mark.parametrize`:
```
    td = TensorDict({}, [10])
    td = TensorDict({"a": torch.zeros(10)}, [10])
    td = TensorDict({"a": torch.zeros(10, 5, 3), "b": torch.zeros(10, 5, 3, 4)}, [10, 5, 3])
```
and have a test where we do
```
for d in range(td.batch_dims):
    rep = td.shape[d] // 2
    split_list = [2] * rep + [td.shape[d] % 2]  % should return [2, 1] for size 3 and [2, 2, 1] for size 5
    for splits in (2, split_list):
        tds = td.split(splits, d)
```

etc

What do you think?

**Review by vmoens — COMMENTED** [review]
Let's have a look at the computation of the batch size :)

File: test/test_tensordict.py (hunk 4 above)
Comment (wonnor-pro) [reply test/test_tensordict.py:2953]: I agree, I'll take the shape of the first tensor up to the batch-size length to ensure the batch size reflects the first few common dimensions.

File: tensordict/tensordict.py (hunk 3 above)
Comment (wonnor-pro) [reply tensordict/tensordict.py:1449]: The `batch_size` and the first dimensions of each of the tensors must be compliant. I previously did not consider this as legal inputs. I'll add the edge cases to the split method. On a side note, can you share more context in how empty tensordict with batch size will be used? Thanks!

File: tensordict/tensordict.py (hunk 2 above)
Comment (wonnor-pro) [reply tensordict/tensordict.py:1436]: I'll update the type format of args `split_size(int or List(int))`.

File: test/test_tensordict.py (hunk 5 above)
Comment (wonnor-pro) [reply test/test_tensordict.py:2938]: Yeah I considered use `pytest.mark.parametrize` as well. 
I think pros are that it will make the test case reflect more of the options we provide in split_size (int or list(int)), but it would also make the test case less expressive and affect the readability, which is important in test files. What do you think?

File: tensordict/tensordict.py (hunk 3 above)
Comment (vmoens) [reply tensordict/tensordict.py:1449]: I would not compute the batch size of the tensordicts based on the size of the tensors
`TensorDict({}, [])` is not splittable
`TensorDict({"a": torch.randn(3)}, [])` is not splittable either (because you can only split along the batch size)
`TensorDict({}, [3, 4])` is splittable (an empty tensordict is like saying "everything you put in this tensordict will have a shape of `[3, 4, *other]`

File: test/test_tensordict.py (hunk 4 above)
Comment (vmoens) [reply test/test_tensordict.py:2953]: Beware of the fact that the batch size is not necessarily the shape up to the last common dimension. It can be anything in between 0 and that.

File: test/test_tensordict.py (hunk 5 above)
Comment (wonnor-pro) [reply test/test_tensordict.py]: As sync-ed up offline, the test function will be moved to TestTensorDict class for wider coverage of test cases. Thanks for the feedback!

File: tensordict/tensordict.py (hunk 3 above)
Comment (wonnor-pro) [reply tensordict/tensordict.py]: The support to empty TensorDict is now implemented. Thanks for the feedback!

**Comment by wonnor-pro** [comment]
I am currently investigating inappropriate behaviour when `dim` is negative in some cases.

**Comment by wonnor-pro** [comment]
The inappropriate behaviours for negative `dim` have been solved.

**Review by vmoens — APPROVED** [review]
Great work!
